### DISCUSSION TITLE
Cannot get staging/preview env file to work

### DISCUSSION THREAD

**Discussion by caweidmann** [vercel/vercel#8719] [Help] [answered]
I have a project setup with Vercel and I am using GitHub actions to deploy the project (copy/paste from [this guide](https://vercel.com/guides/how-can-i-use-github-actions-with-vercel)).

Locally I have the following three env files:
```
// .env.local
NEXT_PUBLIC_FIREBASE_PROJECT_ID=project1234_staging
```
```
// .env.preview
NEXT_PUBLIC_FIREBASE_PROJECT_ID=project1234_staging
```
```
// .env.production
NEXT_PUBLIC_FIREBASE_PROJECT_ID=project1234_PRODUCTION
```

In Vercel UI under the URL `settings/environment-variables` I have the following environment variables declared:
```
// Preview, Development
FIREBASE_PRIVATE_KEY=STAGING_KEY
FIREBASE_CLIENT_EMAIL=STAGING_KEY
```
```
// Production
FIREBASE_PRIVATE_KEY=PROD_KEY
FIREBASE_CLIENT_EMAIL=PROD_KEY
```

Whenever I run a "preview" build and deployment my API routes do not work (they fail to connect to Firebase). I think this is because somehow the app ends up using the `.env.production` file's project ID instead of the `.env.preview` file's project ID - and since I am pulling the "preview" env vars from Vercel via `vercel pull --yes --environment=preview` it now has a mismatch between the project ID and the project's private key and client email. Not sure what I'm missing, but I've been trying to debug this for the past several hours.

If I replace the _values_ in the `.env.production` file with the values from `.env.preview` file then everything works as expected for the "preview" build and deployment. However, if I then push to production my "production" build and deployment is mismatched.

I'm trying to understand why Vercel won't read the `.env.preview` file or what I am doing wrong. Is it even possible to define a `.env.preview` file locally and have Vercel read those values for the "preview" build and deployment?

**Comment by amyegan** [comment]
Hi @caweidmann! Because you're setting the variables in two places, one version of the file is probably overwriting the other. It's best to set all of your env vars in the Project Settings [Environment Variables UI](https://nextjs.org/docs/basic-features/environment-variables#environment-variables-on-vercel). You can [use the CLI](https://vercel.com/docs/cli/env#) to sync those environment variables. The `vercel env pull .env.local` will download Development Environment Variables from the cloud and write them to your `.env.local` for development.

**Reply by caweidmann** [reply]
Hi @amyegan

Thanks for getting back to me. I think, if I'm understanding the CLI correctly, the issue I'm running into is that `vercel build` always references the `.env.production` file. Do you know if there is a way to tell Vercel CLI to use a different `.env` file for the build? E.g. these are the steps I'm using:

```
// (1)
- name: Vercel - Install CLI
  run: npm install --global vercel@latest
// (2)
- name: Vercel - Pull environment information
  run: vercel pull --yes --environment=preview --token=${{ secrets.VERCEL_TOKEN }}
// (3)
- name: Vercel - Build project artifacts
  run: vercel build --token=${{ secrets.VERCEL_TOKEN }}
// (4)
- name: Vercel - Deploy project artifacts
  run: vercel deploy --prebuilt --token=${{ secrets.VERCEL_TOKEN }}
```
I would love to tell step (3) that it addition to using the env vars which were pulled from Vercel UI's "preview" variables, that it should also use the local `.env.preview` file for the build instead of `.env.production`.

One way to solve the problem is to move all the env variables to the Vercel UI. But it feels like a workaround solution as then I'll end up with a lot of variables in the Vercel UI which are much easier to manage locally in the different `.env` files.

**Reply by amyegan** [reply]
Preview is the default and you must specify a production build using the `--prod` option for `vercel build` to use production settings for the build. The [Environment Variable Load Order](https://nextjs.org/docs/basic-features/environment-variables#environment-variable-load-order) page reveals a bit more of what seems to be going wrong in your case. 

I believe `NODE_ENV` will return `production` for both Production _and_ Preview environments because `NODE_ENV` can only reliably be `production`, `development` or `test`. Meaning `.env.preview` is never used in the build step and `.env.production` is found instead. 

I'm going to ask the rest of the team for more info about this. Someone else will know more than I do about how source code environment variables are used during build. Hopefully they will have a good suggestion 🤞

**Reply by caweidmann** [reply]
Thank you! Please do. In the meantime I'll try and play around with NODE_ENV in the build step.

**Reply by caweidmann** [reply]
So I've tried both:
```
- name: Vercel - Build project artifacts
  env:
    NODE_ENV: development
  run: echo "node env = $NODE_ENV" && vercel build --token=${{ secrets.VERCEL_TOKEN }}
```
and 
```
- name: Vercel - Build project artifacts
  env:
    NODE_ENV: preview
  run: echo "node env = $NODE_ENV" && vercel build --token=${{ secrets.VERCEL_TOKEN }}
```
and both times the build loads:
<img width="631" alt="Screenshot 2022-10-13 at 18 51 51" src="https://user-images.githubusercontent.com/1684894/195670375-89ab581a-7388-44bb-8741-da913f34b57d.png">
🤷‍♂️

**Reply by caweidmann** [reply]
I've also tried to use a custom build command with a [project configuration file](https://vercel.com/docs/project-configuration#project-configuration/build-command) as follows:
```
// vercel-preview.json
{
  "buildCommand": "env-cmd -f .env.preview next build"
}
```
and then running:
```
- name: Vercel - Build project artifacts
  run: vercel build --local-config vercel-preview.json --debug --token=${{ secrets.VERCEL_TOKEN }}
```
with the same result. It keeps loading the `.env.production` file.

Here's a trimmed debug log from the GitHub action:
```
Run vercel build --local-config vercel-preview.json --debug --token=***
> [debug] [2022-10-13T18:40:27.736Z] Found config in provided --local-config path /home/runner/work/<project>/vercel-preview.json
Vercel CLI 28.4.10
> [debug] [2022-10-13T18:40:27.746Z] user supplied known subcommand: "build"
> [debug] [2022-10-13T18:40:27.877Z] Loaded environment variables from "/home/runner/work/<project>/.vercel/.env.preview.local"
> [debug] [2022-10-13T18:40:28.068Z] Imported Builder "@vercel/next" from "/opt/hostedtoolcache/node/18.9.0/x64/lib/node_modules/vercel/node_modules/@vercel/next"
> [debug] [2022-10-13T18:40:28.099Z] Building entrypoint "package.json" with "@vercel/next"
WARNING: You should not upload the `.next` directory.
Installing dependencies...

up to date in 2s

115 packages are looking for funding
  run `npm fund` for details
Detected Next.js version: 12.3.1
Detected `package-lock.json` generated by npm 7+...
Running "npm run build"

> <project>@0.4.0 build
> next build

info  - Loaded env from /home/runner/work/<project>/.env.production
```

**Reply by amyegan** [reply]
Sorry for the delayed response. Thank you for providing all of this detail and for your patience! I've asked the wider team for their input and am awaiting feedback.

**Reply by ViliamKopecky** [reply]
There are more troubles with .env files, I believe this is part of it. I prepared a repo that shows whats wrong:

https://github.com/ViliamKopecky/bugreport-vercel-missing-env-variables

// cc @motdotla

**Reply by cglacet** [reply]
>  It's best to set all of your env vars in the Project Settings

I strongly disagree with this, unless there is no other solution (your secrets in an open source project for example) it make no sense to amputate a project from its requirements. Puting environement variables in a 3rd party interface makes you more dependent to the 3rd party which is always a bad news for your project (even if you don't plan on leaving any time soon). 

If you can define environement variables using plain text and your build commands/pipelines, I think its a much better solution.

**Reply by motdotla** [reply]
As @ViliamKopecky demonstrates, this is an inconsistency with how Vercel handles environment variables. It doesn't follow the principle of least surprise.

We recommend adopting the new `.env.vault` standard to fix this issue - until Vercel fixes this.

**Comment by smaeda-ks** [comment, marked as answer]
Hi @caweidmann, just to be clear, Next.js follows this order of reading `.env` files:
https://nextjs.org/docs/basic-features/environment-variables#environment-variable-load-order

and as @amyegan said, with Next.js, `NODE_ENV` would always be `production` regardless of your deployment target (preview vs production) as it's overwritten, except for the local dev server. This is by design, and thus `.env.production` file is always getting loaded when you deploy the app on Vercel, as you saw.

Some context: https://github.com/vercel/next.js/discussions/25764

Some workarounds are:
- Migrate your file-based environment variables to Vercel's native environment variables feature.
  - This adds a few more values, such as [shared environment variable](https://vercel.com/docs/concepts/projects/shared-environment-variables), [bulk upload support](https://vercel.com/changelog/bulk-upload-now-available-for-environment-variables), etc.
- Move your `.env.*` files to a different sub-directory and explicitly load them, so they won't be picked up automatically when building the app. For instance, you can use our `VERCEL_ENV` [system environment variable](https://vercel.com/docs/concepts/projects/environment-variables#system-environment-variables) to reliably tell which environment the build is currently running on. E.g., `env-cmd -f ./path-to-directory/.env.${VERCEL_ENV} next build`

**Reply by cglacet** [reply]
I'm working on a react-native(-web) project where my environement files are located in the parent directory (contains `ios`/`android`/`web` directories). The nextjs app lives in the `web` directory. 

I want three things: 1) bind dot-env filenames that already exist (`.env.dev`, `.env.production`, `.env.staging`) to filenames that match vercel conventions (mostly rename staging to preview). 2) load the preview environment instead of production one when in review, and most importantly 3) I only want a single dot-env file for my native/web apps.

It's a bit clunky (doesn't work on windows local dev), but base on your idea @smaeda-ks, I ended up with this:

```json
{
  "private": true,
  "scripts": {
    "dev": "npm run copy:dot-env &&  env-cmd -f .env.config/.env.${VERCEL_ENV:-development} next",
    "build": "npm run copy:dot-env && env-cmd -f .env.config/.env.${VERCEL_ENV:-development} next build",
    "start": "npm run copy:dot-env && env-cmd -f .env.config/.env.${VERCEL_ENV:-development} next start",
    "copy:dot-env": "cp ../.env.dev .env.config/.env.development && cp ../.env.staging .env.config/.env.preview && cp ../.env.production .env.config/.env.production"
  },
  "devDependencies": {
    "env-cmd": "^10.1.0",
  }
}
```

I added a fallback to `.env.development` because I don't have `VERCEL_ENV` defined locally and I thought it made more sense  to do this instead of defining the variable (easier to setup).

With this solution I think I'll be able to define env variables that are only relative to the web app (in plain `/web/.env.x` files) without intefering with existing environements.

**Reply by ZeldOcarina** [reply]
Oh I got it no local .env file if you are trying to pull envs from the API!

**Reply by jeanbmar** [reply]
The work around proposed by [smaeda-ks](https://github.com/smaeda-ks) isn't great because package.json scripts are a mess and it isn't portable (hello Windows).
It's probably better to inject Next config as explained here: https://github.com/vercel/next.js/discussions/25764

**Reply by khanakia** [reply]
Getting error when i run the build command locally
```
➜  dash git:(main) ✗ yarn build                                      
yarn run v1.22.19
$ env-cmd -f ./.env.${VERCEL_ENV} next build
Error: Failed to find .env file at path: ./.env.
    at getEnvFile (/Users/diana/D1/www/projects/test/hellojs/node_modules/env-cmd/dist/get-env-vars.js:40:19)
```

**Reply by ViliamKopecky** [reply]
> Getting error when i run the build command locally
> 
> ```
> ➜  dash git:(main) ✗ yarn build                                      
> yarn run v1.22.19
> $ env-cmd -f ./.env.${VERCEL_ENV} next build
> Error: Failed to find .env file at path: ./.env.
>     at getEnvFile (/Users/diana/D1/www/projects/test/hellojs/node_modules/env-cmd/dist/get-env-vars.js:40:19)
> ```

You probably dont have VERCEL_ENV available on your local machine. You can configure fallback to `development` like this `${VERCEL_ENV:-development}`

**Reply by khanakia** [reply]
Yes, it is not isn't `vercel build` command should automatically populate it otherwise how i will make the solution suggested by @smaeda-ks. 

It seems in order to make it work now i have to pass VERCEL_ENV always on local which adds another unnecessary complexity

**Reply by motdotla** [reply]
We've run into this as well. We've moved to using a .env.vault file. It's early days but it drastically simplifies this environment handling in production for frameworks like Vercel.

You can read more here: https://github.com/motdotla/dotenv#-deploying

**Reply by khanakia** [reply]
@motdotla I have checked that but how will it resolve that `VERCEL_ENV` issue also next.js has its own internal functions to parse the `.env.*` files so i am unable be able to replace next.js with `dotenv` package

**Reply by motdotla** [reply]
We'll be releasing `16.1.0` soon and then Next.js should soon upgrade to the latest dotenv lib - then this should all be available by default. Until then, keep lobbying them to simplify this.

**Reply by ViliamKopecky** [reply]
@motdotla have you tested it with edge runtime and middlewares on vercel? In my experience - these sometimes get only env vars that are configured through vercel dashboard.

**Reply by motdotla** [reply]
@ViliamKopecky working on this today and will add it to the dotenv examples at https://github.com/motdotla/dotenv#-examples

**Reply by motdotla** [reply]
I put together a blog post on this:

https://www.dotenv.org/blog/2023/05/22/use-dotenv-vault-with-nextjs.html

This uses the `.env.vault` mechanism and is decrypted on build and deploy.

## Additional Information

Edge functions use web-native libraries only. You cannot use `fs`, `process.cwd` etc.
Instead, they are pre-built and deployed. As far as I can tell, their source code is not exposed to source like the front-end code of a NextJS app.

Therefore, I think it is safe to compile these edge functions with `NEXT_PUBLIC_` envs. I recommend prefacing them with `NEXT_PUBLIC_EDGE_` to separate them in your mind. Also, as far as I know, NextJS long ago fixed the problem of including NEXT_PUBLIC_ envs in the build if they were not referenced in code.

A final note, it appears to me that Vercel makes a special case for its own environment variables. That is why you don't have to preface theirs with `NEXT_PUBLIC_` for the edge functions. My take is that it is inconsistent behavior and an oversight on their part.

@ViliamKopecky what are your thoughts?

**Reply by ViliamKopecky** [reply]
@motdotla Very nice... I dont know if I can get over the `NEXT_PUBLIC_` prefix - it just seems so risky as I can imagine Vercel could distribute those globally in plaintext (altough I hopo not).

Another weird workround in my mind would be setting up env with Vercel REST API... which has also problems:

- not transactional: if you change env before build and build fails - env should be reverted to the state before. This could be implemented with custom build script, just not ideal.
- you would need Vercel API key... which has all the permissions as your user and cannot be currently scoped. Some scoping is available on enterprise tariff, and even then I dont believe you could have something as specific as `allow manage env variables for project X`.

It is Vercel's bug - I reported it in december and they said they will look into it... then later in february they said they submitted a feature request but dont have an ETA.
